dockstore-client/src/main/java/io/dockstore/client/cli/nested/BaseLanguageClient.java

```diff
@@ -0,0 +1,305 @@
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import com.google.common.io.Files;
+import io.dockstore.client.cli.nested.NotificationsClients.NotificationsClient;
+import io.dockstore.common.Utilities;
+import io.swagger.client.ApiException;
+import io.swagger.client.model.ToolDescriptor;
+import org.apache.commons.configuration2.INIConfiguration;
+import org.apache.commons.exec.ExecuteException;
+import org.apache.commons.lang3.tuple.MutableTriple;
+import org.apache.commons.lang3.tuple.Triple;
+
+import static io.dockstore.client.cli.ArgumentUtility.errorMessage;
+import static io.dockstore.client.cli.ArgumentUtility.exceptionMessage;
+import static io.dockstore.client.cli.ArgumentUtility.out;
+import static io.dockstore.client.cli.Client.API_ERROR;
+import static io.dockstore.client.cli.Client.ENTRY_NOT_FOUND;
+import static io.dockstore.client.cli.Client.GENERIC_ERROR;
+import static io.dockstore.client.cli.Client.IO_ERROR;
+
+/**
+ * A base class for all language clients
+ * Clients for CWL, WDL, Nextflow, etc should extend this and implement the abstract functions.
+ */
+public abstract class BaseLanguageClient {
+    protected final AbstractEntryClient abstractEntryClient;
+    protected INIConfiguration config;
+    protected String notificationsWebHookURL;
+    protected NotificationsClient notificationsClient;
+    protected BaseLauncher launcher;
+
+    // Fields set on initial launch
+    protected boolean isLocalEntry;
+    protected String entry;
+    protected String wdlOutputTarget;
+    protected String yamlParameterFile;
+    protected String jsonParameterFile;
+    protected String uuid;
+
+    // Fields generated during setup and running of entry
+    protected File tempLaunchDirectory;
+    protected File localPrimaryDescriptorFile;
+    protected File importsZipFile;
+    protected String selectedParameterFile;
+    protected File provisionedParameterFile;
+    protected String workingDirectory;
+    protected String stdout;
+    protected String stderr;
+
+    public BaseLanguageClient(AbstractEntryClient abstractEntryClient, BaseLauncher launcher) {
+        this.abstractEntryClient = abstractEntryClient;
+        this.launcher = launcher;
+    }
+
+    public void setLauncher(BaseLauncher launcher) {
+        this.launcher = launcher;
+    }
+
+    /**
+     * Selects the intended parameter file
+     */
+    public abstract String selectParameterFile();
+
+    /**
+     * Provision the input files based on the selected parameter file.
+     * Creates an updated version of the parameter file with new local file locations.
+     * Must set the variable workingDirectory
+     * @return Updated parameter file
+     */
+    public abstract File provisionInputFiles();
+
+    /**
+     * Runs the tool/workflow with the selected launcher
+     * Must set the variables stdout and stderr
+     * @throws ExecuteException
+     */
+    public abstract void executeEntry() throws ExecuteException;
```

Comment:
Could executeEntry() perhaps be a method in BaseLauncher? I was thinking that we might want a WESLauncher like CromwellLauncher or CWLToolLauncher. Right now each ...LanguageClient implements executeEntry as a local command to run, but it might be good to delegate that to the Launcher so it could in the case of WESLauncher make a REST request instead. I haven't thought though it completely so I'm not sure if there are negative side effects.

Reply:
I think that is a good idea. I'll make the change.